api/agent/runner_client.go

```diff
@@ -90,120 +100,191 @@ func (r *gRPCRunner) TryExec(ctx context.Context, call pool.RunnerCall) (bool, e
 		// If we can't encode the model, no runner will ever be able to run this. Give up.
 		return true, err
 	}
+
 	runnerConnection, err := r.client.Engage(ctx)
 	if err != nil {
 		logrus.WithError(err).Error("Unable to create client to runner node")
 		// Try on next runner
 		return false, err
 	}
 
+	// After this point, we assume "COMMITTED" unless pure runner
+	// send explicit NACK
 	err = runnerConnection.Send(&pb.ClientMsg{Body: &pb.ClientMsg_Try{Try: &pb.TryCall{ModelsCallJson: string(modelJSON)}}})
 	if err != nil {
 		logrus.WithError(err).Error("Failed to send message to runner node")
-		return false, err
-	}
-	msg, err := runnerConnection.Recv()
-	if err != nil {
-		logrus.WithError(err).Error("Failed to receive first message from runner node")
-		return false, err
+		return true, err
```

Comment:
Do we need to do anything to close the runnerConnection on error?

Reply:
I don't think so, corresponding generated code intentionally leaks the stream (runner.pb.go Engage()) and returns nil. And we'll be cancelling this context also. We should be OK.